### ISSUE TITLE
DateSerializer synchronizes on _customFormat

### ISSUE THREAD

**Issue by brenuart** [FasterXML/jackson-databind#1652]
The `DateSerializer#serialize(...)` method *always* synchronizes on the custom format when one is specified (see https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/com/fasterxml/jackson/databind/ser/std/DateSerializer.java#L49-L53).

However, it seems a custom format is specified only in the following cases:
- when a _contextual_ DateSerializer instance is created (in the `DateTimeSerializeBase#createContextual(...)` method. In this case, the custom format is "one shot" and there is no need to protect it from concurrent accesses (synchronisation is not required)
- when someone creates an instance of DateSerializer with its own date format instance. In this case you don't know if it is thread safe or not..

For case (1), it means that usage of the `@JsonFormat` annotation will necessarily go through the synchronised path and impose potentially serious performance penalty. 

For case (2), a first step might be to refactor the `serialize()` method and split it into `serializeAsTimestamp()`, `serializeCustom()` and `serializeDefault()`... At least we would have been able to override the _custom_ version and add/remove the synchronisation... ;-)

PS: I wouldn't call this issue a _bug_... This is more sharing opinions and/or asking for advices ;-)

**Comment by cowtowncoder** [comment]
I would counter that for this to be considered an issue, there should be some supporting evidence.
And given that:

1. Date processing is hideously inefficient process in general (try measuring it for one), and
2. Uncontested synchronization has very low overhead

I think that although I concur that theoretically one can figure out what cases require synchronization (ones where no `clone()` or equivalent it used to create non-sharde instance), it may not be worth the hassle. Especially since missing just one case where sync was needed for a shared `DateFormat` could cause hard to diagnose issues and data corruption.

**Comment by brenuart** [comment]
I agree with you except for case (1) mentioned above. Our application dropped from `1.000.000 parsing/s` to `200.000 parsing/s` just because we added a `@JsonFormat` annotation on a field :(
This application is obviously multithreaded...

**Comment by cowtowncoder** [comment]
@brenuart that would certainly be measurable difference, and if it can be shown that synchronized block is to blame (and that eliminating it does not cause concurrency fails) it would warrant work.

One other thing to note on sharing is this: serializers and deserializers are shared and need to be stateless (similar to how, say, `Servlet`s are). So there is real need to guard concurrent usage somehow: either by synchronized block, or by cloning `DateFormat` instance, at least across threads.

**Comment by brenuart** [comment]
I created a sample test to show how the `@JsonFormat` annotation could affect serialisation performances in a multi-threaded scenario. The code is available at https://gist.github.com/brenuart/191ae24dd9f722140e8ce5c8052df0e2

The test serialises two different POJOs, with and without annotation on its single field. It runs for 5 seconds and reports the serialisation rate per thread. The test is repeated with 1, 2, 3 and 4 concurrent threads to  measure the potential impact of concurrency.

In an ideal world the rate per thread would remain the same, what ever the number of threads (as long as it is not higher than the number of cpu cores of course). It would however drop significantly if there is some contention between threads. Below are the numbers I see when running the test on my machine:

```
PojoPlain, Thread count: 1, Rate per thread: 790351 write/thread/s
PojoPlain, Thread count: 2, Rate per thread: 897603 write/thread/s
PojoPlain, Thread count: 3, Rate per thread: 853100 write/thread/s
PojoPlain, Thread count: 4, Rate per thread: 816039 write/thread/s

PojoAnnot, Thread count: 1, Rate per thread: 1281769 write/thread/s
PojoAnnot, Thread count: 2, Rate per thread: 612441 write/thread/s
PojoAnnot, Thread count: 3, Rate per thread: 402384 write/thread/s
PojoAnnot, Thread count: 4, Rate per thread: 300590 write/thread/s
```

`PojoPlain` is the version without annotations. As you can see the rate is pretty stable despite the increased number of threads. `PojoAnnot` is the version annotated with `@JsonFormat(timezone="GMT+4")`. As you can see, the rate with two threads is half the single thread. Three threads is 1/3 of with a single thread...
The presence of the annotation on the field has clearly a negative impact on the performances in a multi-thread scenario.

The test does a third round with an _unsynchronized_ version of the `DateSerializer`. This implementation is pretty naive and its only purpose is to highlight the impact of the synchronised block. Here are the numbers for the annotated pojo:

```
PojoAnnot, Thread count: 1, Rate per thread: 1421521 write/thread/s
PojoAnnot, Thread count: 2, Rate per thread: 1161116 write/thread/s
PojoAnnot, Thread count: 3, Rate per thread: 1042114 write/thread/s
PojoAnnot, Thread count: 4, Rate per thread: 997214 write/thread/s
```

Looks better. I don't know why numbers are getting lower as the number of threads increases however (GC ?)...

**Comment by cowtowncoder** [comment]
Thank you. Regarding throughput: number of cpu cores should be limiting factor.

The difference wrt non-annotated case is probably this: default `Date` serialization uses method defined in `SerializerProvider`, which eventually calls:

```java
    protected final DateFormat _dateFormat()
    {
        if (_dateFormat != null) {
            return _dateFormat;
        }
        DateFormat df = _config.getDateFormat();
        _dateFormat = df = (DateFormat) df.clone();
        return df;
    }
```

so that essentially `DateFormat` is clone()d once (and only once) per serialization operation (`writeValue()`). It does not block on synchronized because each thread gets its own copy.

Unfortunately this can only be used for default operations, since `SerializerProvider` and `DeserializationContext` instances are created for each operation, and are thread-specific (that is, instances not shared across threads).
Serializers and deserializers, on the other hand, are shared. So they must either create new `clone()`, or synchronize shared instance.

Now: one thing that might be true is that while there is state `SimpleDateFormat` contains during reading (`parse(...)`), perhaps it is not the case for writing (`format(...)`).
But javadocs of `DateFormat` simply state:

```
Date formats are not synchronized. It is recommended to create separate format instances for each thread. If multiple threads access a format concurrently, it must be synchronized externally.
```

So I don't think this can be counted on.

Now... whether clone()ing for each field read/write (in case of non-default format) is more efficient than synchronizing really depends on couple of things:

1. Cost of clone()ing
2. Ratio of time spend on reading/writing `Date`s for given field

because real time penalty here is -- I think -- due to real collision, multiple threads writing contents of same field. And this ratio likely varies a lot, and is highest on simplest test cases (like POJO with just one `Date` field).

What a mess. :)

Anyway: one thing I would recommend (but which may not be practical) would be to consider use of either `Joda` or Java 8 date/time types. Their thread-safety handling is simpler and should not add similar overhead, because formatter objects have well-defined contracts. There are other benefits too.

**Comment by cowtowncoder** [comment]
Also: if it wasn't clear from above, I am not 100% how to proceed with this issue. I concur that synchronization can be problematic, but am not sure how it could be avoided.

**Comment by cowtowncoder** [comment]
No plans to work on this one, closing.
